### PR TITLE
WMR soft limit

### PR DESCRIPTION
**Pull request by ggevay**
This PR implements the WMR soft limit in the form that we agreed on with @aalexandrov. (I'll update the [design doc](https://github.com/MaterializeInc/materialize/pull/18538) tomorrow.) The "soft" means that we don't error out when we reach the limit, but just stop iterating, and consider the current state as the final result.

**The default limit is infinite, i.e., no limit.**

We postponed the hard limit (erroring out when reaching the limit), because we now have proper dataflow cancellation for WMR queries. We can still [consider](https://github.com/MaterializeInc/database-issues/issues/5564) a hard limit later (which would be easy to add after this PR).

We moved away from system/session variables, based on feedback on the [design doc](https://github.com/MaterializeInc/materialize/pull/18538) from Surfaces and Frank.

Instead of system/session variables, users can specify the limit using new SQL syntax (**Edit:** I'm in the process of changing this to a more standard options clause (but leave it at the same place)):
```WITH MUTUALLY RECURSIVE MAXITERATIONS 42```
This syntax allows for separate limits for each WMR block, which is [important](https://github.com/ggevay/materialize/blob/wmr-limit-design-doc/doc/developer/design/20230330_recursion_limit.md#queries-with-multiple-wmr-clauses----usability) when a query has multiple WMR blocks.

The first commit is just changing iteration numbers to use `u64` instead of `usize`, based on a discussion in the office hours. The second commit runs `cargo fmt`, which I will squash into the first one, of course. (Just wanted to separate the diff for review.)

The second commit is the meat.

@aalexandrov, is the EXPLAIN format ok? (HIR, MIR, LIR) Note that I'm not currently testing `linear_chains`, because it [doesn't seem to work for WMR (and nobody seems to be using it)](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1682451462451629). We can discuss whether to fix it or deprecate it. (I'll open an issue tomorrow.)

#### Keyword

The keyword is tentative; we can still decide that before merging, but I like `MAXITERATIONS`. (@ggnall) Some possible alternatives:
- `MAXDEPTH 42`: users might think of recursive function calls, which is not what's happening here.
- `LIMIT 42`: It's not clear what are we limiting: iterations, records, time, ... Also, it could be confused with ORDER BYs LIMIT.
- `(ITERATE 42 TIMES)`: This one would also be ok I guess, but I'd vote for simplicity.
- `RECURSIONLIMIT`: Might be ok as well, but I wanted to avoid (internal) confusion with our `pub const RECURSION_LIMIT`, which is for something totally different.
- `MAXRECURSION`: Edit: The MySQL option with the same keyword would correspond to a _hard_ limit, so let's not align our soft limit's keyword with that.
- `ITERATIONS`: ([from Jan](https://github.com/MaterializeInc/materialize/pull/18538#discussion_r1189689052)) We don't have to call it a limit, since running to an exactly specified number of iterations is the same as stopping when either a limit is reached or fixpoint is reached. But I'm a bit worried that some users won't make this extra mental step of realizing that it might execute less iterations if a fixpoint is reached earlier, and might get worried that it will take more steps than necessary.

### Motivation

  * This PR adds a known-desirable feature: https://github.com/MaterializeInc/database-issues/issues/5409

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated design doc: https://github.com/MaterializeInc/materialize/pull/18538
  - (The design doc currently doesn't reflect the move away from session/system variables as well as the hard limit's importance going down with the dataflow shutdown. I'll update the design doc a bit later.)
  
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
  - We do have a protobuf change for the LIR `LetRec`, but I asked Jan whether we need to be backwards-compatible, and he said "Nope, we are not durably storing serialized IRs anywhere, I’m pretty sure. So no need to be backwards-compatible in the protobufs"
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/sql-parser/src/ast/defs/query.rs
Comment (mjibson) [inline src/sql-parser/src/ast/defs/query.rs:367]: This needs a test in https://github.com/MaterializeInc/materialize/blob/main/src/sql-parser/tests/testdata/select

Comment (ggevay) [reply src/sql-parser/src/ast/defs/query.rs:367]: Thanks! Just saw your msg on Slack. I'll definitely add a test.

(I'll ping you when the PR is ready for review.)

Comment (ggevay) [reply src/sql-parser/src/ast/defs/query.rs:373]: Done.

**Comment by ggevay** [comment]
Ready for review! @MaterializeInc/surfaces, @aalexandrov, @vmarcos (rendering).

File: src/sql-parser/src/ast/defs/query.rs
Comment (mjibson) [inline src/sql-parser/src/ast/defs/query.rs:347]: As an aside (don't need to do anything in this PR): having logic in the parser is unfortunate and it'd be nice for it to not be here.

File: src/sql-parser/src/parser.rs
Comment (mjibson) [inline src/sql-parser/src/parser.rs:4636]: This logic should live in the planner, not the parser.

File: src/compute/src/render/mod.rs
Comment (teskje) [inline src/compute/src/render/mod.rs:103]: Nit: This should be in the third-party import block below.

File: src/compute/src/render/mod.rs
Comment (teskje) [inline src/compute/src/render/mod.rs:663]: Nit: `if let Some(max_iter) = max_iter { ... }` would be slightly more readable.

**Review by philip-stoev — APPROVED** [review]
Thank you very much for the tests -- I could not figure out any additional ones to add.

The feature held under manual experimentation and I could not get it to become wedged .

The diff looks less formidable with whitespace ignored.

File: src/compute-client/src/plan.proto
Comment (teskje) [inline src/compute-client/src/plan.proto:80]: I assume there is no `repeated optional uint64` that would make this second field unnecessary?

Relatedly, but not really relevant to this PR, I have been wondering why `LetRec` is represented with separate lists for ids and values (and now max_iters). ISTM that instead having a single `bindings` list with `(id, value, max_iter)` entries would be more convenient since it doesn't require ensuring that the lists have the same length all the time. But I'm probably missing something.

File: src/expr/src/explain/text.rs
Comment (teskje) [inline src/expr/src/explain/text.rs:310]: I'm not a huge fan of making our code specific to the limitations of specific IDEs!

I can't think of a good solution, but it makes my wonder why we are not running into this issue more often. Is this the only instance where CLion is not able to able to handle code that rustc is, or do you usually ignore errors? I assume more people are not using CLion, so it must happen from time to time that someone checks in code that's not compatible with it.

File: src/expr/src/explain/text.rs
Comment (teskje) [inline src/expr/src/explain/text.rs:330]: This looks like it can be a single `reduce`:
```rust
let all_max_iters_same = max_iter.iter().reduce(|first, i| {
    if i == first { first } else { &None }
});
```

File: src/transform/src/normalize_lets.rs
Comment (teskje) [inline src/transform/src/normalize_lets.rs:164]: Nit: `use support::replace_bindings_from_map`.

**Review by teskje — APPROVED** [review]
General compute parts LGTM, I just have a couple stylistic comments. I'll defer to @aalexandrov for the transform changes.

**Review by vmarcos — APPROVED** [review]
Rendering changes look good to me; it would be ideal if we'd include test(s) for nested WMR blocks with different limits also in execution (`test/sqllogictest/with_mutually_recursive.slt`), though, not only in planning.

File: src/compute-client/src/plan.proto
Comment (aalexandrov) [reply src/compute-client/src/plan.proto:80]: We discussed that this might be a better representation but decided to not change it as part of the ongoing epic. I think it must be a skunkworks project or something similar, unless we find ~3-4 days of in-band work to do this as part of addressing tech debt.

File: test/sqllogictest/explain/physical_plan_as_text.slt
Comment (aalexandrov) [reply test/sqllogictest/explain/physical_plan_as_text.slt:882]: nit: stylistically parameters of AST nodes have been printed as `max_iterations` elsewhere, so I think
```text
cte [max_iterations=10] l1 =
```
is a bit more consistent with the rest of the format.

**Comment by aalexandrov** [comment]
One general comment: we should absolutely prohibit people setting the MAX_ITERATIONS for some binding to `0`. There are some `NonZero~` types in https://doc.rust-lang.org/stable/std/num/index.html and I guess the most precise way to enforce this constraint is to use one of those.

File: src/compute-client/src/plan.proto
Comment (teskje) [reply src/compute-client/src/plan.proto:80]: Makes sense, thanks!

Regarding `max_iters_present`, Alex mentioned in another comment:
> One general comment: we should absolutely prohibit people setting the MAX_ITERATIONS for some binding to 0, otherwise we run the risk of all sorts of incorrect optimizations for those blocks (or unnecessary complicated transform code).

Doing that would allow us to drop `max_iters_present` here and just use 0 to mean "no limit set".

File: src/compute/src/render/mod.rs
Comment (ggevay) [reply src/compute/src/render/mod.rs]: Yes, thanks, changed it.

File: src/compute/src/render/mod.rs
Comment (ggevay) [reply src/compute/src/render/mod.rs]: The import block below is DD/Timely. I created a new import block between std and the DD/Timely block. Is this what you meant?

File: src/compute-client/src/plan/mod.rs
Comment (ggevay) [reply src/compute-client/src/plan/mod.rs]: Fixed

File: src/compute-client/src/explain/text.rs
Comment (ggevay) [reply src/compute-client/src/explain/text.rs]: Fixed

File: src/compute-client/src/plan.proto
Comment (teskje) [reply src/compute-client/src/plan.proto:80]: I'd prefer it, since it statically removes the possibility that the two lists might become inconsistent (e.g. have different lengths), so there is one thing less to check at runtime. But your argument about special values is valid too, so I don't oppose leaving things as they are.

File: src/compute/src/render/mod.rs
Comment (teskje) [reply src/compute/src/render/mod.rs]: Normally (though I don't think that's spelled out anywhere), we structure our imports into blocks like this:

* `std` imports
* third-party imports
* `mz_` imports
* `crate` imports
* `super` imports
* `self` imports

(The last three blocks I'm actually not sure about. I think they are sometimes merged into one.)

According to this, DD, Timely, and `itertools` all should be in the third-party import block. I don't think we usually do dedicated import blocks for DD/Timely. In this case it just looks like we do because those are the only third-party crates we use in this module.

File: src/expr/src/explain/text.rs
Comment (ggevay) [reply src/expr/src/explain/text.rs]: Yes, there are some situations where it can't figure out some types. Here, I think the `izip` macro trips it up. There are also some situations where it detects a type error and makes a red underlining when in fact there is no type error. But these situations are rare enough that I can just ignore it with negligible impact on my productivity.

I deleted the explicit type annotation.

File: src/expr/src/explain/text.rs
Comment (ggevay) [reply src/expr/src/explain/text.rs]: Thanks!

File: src/transform/src/normalize_lets.rs
Comment (ggevay) [reply src/transform/src/normalize_lets.rs]: Fixed

File: src/compute/src/render/mod.rs
Comment (ggevay) [reply src/compute/src/render/mod.rs]: Ok, fixed, thanks!

File: src/sql-parser/src/parser.rs
Comment (ggevay) [reply src/sql-parser/src/parser.rs]: Ok, I moved it.

**Comment by ggevay** [comment]
Thanks for the reviews! In addition to addressing the inline comments, I made the following changes:

I added the test from @philip-stoev.

I added the test that @vmarcos suggests.

I changed the LIR EXPLAIN to not print it if None, as @aalexandrov suggested.

> we should absolutely prohibit people setting the MAX_ITERATIONS for some binding to 0, otherwise we run the risk of all sorts of incorrect optimizations for those blocks (or unnecessary complicated transform code).

Yes, I'm throwing an error for 0. (while creating the HIR from the AST)

> There are some NonZero~ types

Hmm, nice! I changed the code to use this. (I didn’t properly implement `MzReflect for NonZeroU64`, but I hope that’s ok. We are not planning to use lowertest for WMR I guess, since we are moving away from lowertest anyway. @aalexandrov)

File: src/sql-parser/src/parser.rs
Comment (ggevay) [reply src/sql-parser/src/parser.rs]: The PR should be in it's final form now. Could you please do a final check @mjibson?

**Review by mjibson — APPROVED** [review]

**Comment by ggevay** [comment]
@benesch, you were auto-added as a reviewer due to adding a small thing in mz-ore in `src/ore/src/option.rs`.

**Comment by ggevay** [comment]
Is the keyword and syntax ok? Who has the authority on this? @ggnall? A simple example:
```
WITH MUTUALLY RECURSIVE MAXITERATIONS 3
  cnt (i int) AS (
    SELECT 1 AS i
    UNION
    SELECT i+1 FROM cnt)
SELECT * FROM cnt;
```

**Review by benesch — CHANGES_REQUESTED** [review]
Yes, we need a design proposal for the syntax. A short design document would be ideal, but a comment in a GitHub issue or on this PR would do fine.

The current syntax does not follow any of our established idioms for specifying options or naming parameters. We have the start of a guide for these idioms here: https://www.notion.so/materialize/SQL-design-principles-2e2069797f7c46b8b4c7dd10043afdcd

I suggest that we model this after the existing `OPTIONS` and `WITH` clauses:

```sql
WITH MUTUALLY RECURSIVE (MAX ITERATIONS 1)
   cnt (i int) AS (SELECT ...)
```

Where `MAX ITERATIONS` is two separate tokens, per our style guide, and the bit in the parentheses is a standard options block, where you can also say `MAX ITERATIONS = 3`.

Is "iterations" the right term here? Given that the feature is "with mutually **recursive**", would we prefer `RECURSION LIMIT` to `MAX ITERATIONS`?

File: src/compute-client/src/plan/mod.rs
Comment (aalexandrov) [inline src/compute-client/src/plan/mod.rs:607]: minor nit: can you add this 

```rust
impl RustType<u64> for Option<NonZeroU64> {
    fn into_proto(&self) -> u64 {
        match self {
            Some(d) => d.get(),
            None => 0,
        }
    }

    fn from_proto(proto: u64) -> Result<Self, TryFromProtoError> {
        Ok(NonZeroU64::new(proto))
    }
}
```

after this line

```rust
rust_type_id![bool, f32, f64, i32, i64, String, u32, u64, Vec<u8>];
```

and then use `max_iters.into_proto()` and `proto.max_iters.into_rust()?`?

**Review by aalexandrov — APPROVED** [review]
Looks good, from my original suggestions I think the only thing missing is changing the output format of the new attribute from
```text
[MaxIterations $x]
```
to
```text
[max_iterations=$x]
```

File: src/compute-client/src/plan.proto
Comment (ggevay) [reply src/compute-client/src/plan.proto]: Changed it.

File: test/sqllogictest/explain/physical_plan_as_text.slt
Comment (ggevay) [reply test/sqllogictest/explain/physical_plan_as_text.slt]: Yes, fixed, thx!

**Comment by ggevay** [comment]
I've addressed all comments, including changing the syntax based on the SQL design principles notion doc and SELECT's expected group size option. An example for the syntax:
```
WITH MUTUALLY RECURSIVE (ITERATION LIMIT 6)
  cnt (i int) AS (
    (WITH MUTUALLY RECURSIVE (ITERATION LIMIT = 3)
       cnt (i int) AS (
         SELECT 1 AS i
         UNION
         SELECT i+1 FROM cnt)
       SELECT i FROM cnt
    )
    UNION
    SELECT i+100 FROM cnt)
SELECT i FROM cnt;
```
We can decide the exact keywords later, after we finalize the keywords for WITH MUTUALLY RECURSIVE itself. (One option that came up was WITH REPEATEDLY, which doesn't have RECURSIVE in it, so the word ITERATION would be ok for it.)

@benesch, @ggnall, could you please take a quick look at the syntax?

The surfaces code changed because of the different option parsing. @mjibson, could you please check the new parsing and planning?

I also updated the [design doc](https://github.com/MaterializeInc/materialize/pull/18538).

File: src/compute/src/render/mod.rs
Comment (def-) [inline src/compute/src/render/mod.rs:281]: This looks like an interesting block to test

File: src/compute/src/render/mod.rs
Comment (def-) [inline src/compute/src/render/mod.rs:291]: Interesting block to test

File: src/compute/src/render/mod.rs
Comment (def-) [inline src/compute/src/render/mod.rs:665]: Interesting block to test

File: src/expr/src/explain/text.rs
Comment (def-) [inline src/expr/src/explain/text.rs:323]: This entire block is untested it seems?

**Review by def- — COMMENTED** [review]
Some interesting spots from code coverage report marked inline: https://buildkite.com/materialize/coverage/builds/89

File: src/expr/src/explain/text.rs
Comment (ggevay) [reply src/expr/src/explain/text.rs:323]: Yes, because linear chains currently doesn't work with WMR, see https://github.com/MaterializeInc/materialize/issues/19012.

**Review by mjibson — COMMENTED** [review]
Surfaces parts lgtm

File: src/compute/src/render/mod.rs
Comment (ggevay) [reply src/compute/src/render/mod.rs:665]: Hmm, well, this is definitely covered in `with_mutually_recursive.slt` (by most of the new tests), as this is the core of the PR. I verified this now by putting a crash into this block and running `with_mutually_recursive.slt`.

What tests does coverage take into account? Does it consider slt tests? Or how can it be that it shows it as not covered?

File: src/compute/src/render/mod.rs
Comment (ggevay) [reply src/compute/src/render/mod.rs:291]: This is indeed not covered, thanks! I'll add a test for this. (Although this PR makes only a trivial change to this part.)

File: src/compute/src/render/mod.rs
Comment (ggevay) [reply src/compute/src/render/mod.rs:281]: This is actually already covered by `with_mutually_recursive.slt`.

File: src/compute/src/render/mod.rs
Comment (def-) [reply src/compute/src/render/mod.rs:665]: It should consider slt tests: https://github.com/MaterializeInc/materialize/blob/main/doc/developer/code-coverage.md I'm seeing a problem with the coverage data here, will investigate.

File: src/compute/src/render/mod.rs
Comment (def-) [reply src/compute/src/render/mod.rs:281]: Thanks, you are right, this was an issue in my coverage implementation, fixed with https://github.com/MaterializeInc/materialize/pull/19193

File: src/compute/src/render/mod.rs
Comment (ggevay) [reply src/compute/src/render/mod.rs:291]: Added a test in aebe703f4e3fe72b3982094dc0cb116920148231

**Comment by benesch** [comment]
New syntax looks much better, nice! I signal boosted in #devex (https://materializeinc.slack.com/archives/C015RHB3LDR/p1683779552919369) for additional feedback. "Iteration limit" sounds a little funny to my ear, but no strong feelings.

**Comment by morsapaes** [comment]
> We can decide the exact keywords later, after we finalize the keywords for WITH MUTUALLY RECURSIVE itself. (One option that came up was WITH REPEATEDLY, which doesn't have RECURSIVE in it, so the word ITERATION would be ok for it.)

Any user that is familiar with recursion in SQL will have a mental mapping for `WITH RECURSIVE`, so it doesn't sound right to me to use an entirely different term for it (like `WITH REPEATEDLY`). Are we **forced** to use the `MUTUALLY` keyword, or would it be possible to relax this to `WITH RECURSIVE`, and spell out how we depart from the SQL standard implementation?

For the same reason, my preference for the iteration limit would also be for a keyword that is used in other systems, like `MAXRECURSION` or `MAXDEPTH `. AFAIU, these refer to the maximum depth level, though (which seems to be how most databases limit recursive queries?), so if that's a fundamentally wrong way to think about it, it might be safer to use `MAXITERATIONS`.

**Comment by benesch** [comment]
> From my point of view, simply using `WITH RECURSIVE` sounds ok, but Frank said that Nikhil might not want that. The only potential issue that I can see is that we have a semantic difference, i.e., some recursive queries have different results between Postgres and Materialize.

Yeah, I want to keep the door open for supporting the SQL standard's semantics for `WITH RECURSIVE`. It may be important for a customer one day.

@petrosagg had the take of: why bother with the `RECURSIVE` keyword at all? Just allow CTEs in a normal `WITH` block to refer to one another. I think that's more plausible, because that's a strict extension to the SQL spec, rather than 

> MySQL errors out when reaching MAXRECURSION, which would correspond to our hard limit. In contrast, the soft limit (this PR) simply produces the current state as the final result when reaching the limit. For this reason, I think we shouldn't align the keyword.

Oh, interesting! That makes sense for debugging (you want to watch it progress), but kind of scary if you were running in production. Should we be very explicit with our keyword and include something like `SOFT` in the name? `SOFT ITERATION LIMIT`, for example? Although I'm not sure it's exactly a "soft limit"—I think of a "soft" vs "hard" as "changeable upon request" vs "enforced", whereas this is about "errors or not if limit reached." Naming is hard!

Just noodling:

```
WITH ... (ITERATION LIMIT = 256, REQUIRE FIXPOINT = false)
WITH ... (ITERATION LIMIT = 256, REQUIRE FIXPOINT = true)
WITH ... (ITERATION LIMIT = 256) -- REQUIRE FIXPOINT defaults to true
```

----

I think to avoid blocking this PR any longer we should move forward with either `ITERATION LIMIT` or `RECURSION LIMIT`, but let's sync again on the syntax holistically once this is closer to stabilization.

**Comment by teskje** [comment]
Another idea I had was to just use the keyword `ITERATIONS` without calling it a limit: https://github.com/MaterializeInc/materialize/pull/18538#discussion_r1189689052.

**Comment by ggevay** [comment]
I like Petros' and Jan's simplifying ideas:
- Simply `WITH`, no `RECURSIVE` is needed.
- Simply `ITERATIONS`, no `LIMIT` is needed.

I'm not sure about the [soft vs. hard limit](https://github.com/MaterializeInc/materialize/pull/18966#issuecomment-1544785102). I think if the docs are very explicit about not erroring but simply stopping by default, then it's ok to simply stop by default.

Merging now with ITERATION LIMIT, and then let's decide these things as a follow-up, to already let people use the feature (internally), and also to avoid rebasing this PR again and again.

**Pull request merged**
